### ISSUE TITLE
What are the high level changes that helped improved connection time ?

### ISSUE THREAD

**Issue by darshan-verkada** [aws-samples/kvs-connection-time-optimization-for-webrtc-sample#1]
**TL;DR**: Is it the suggestion that to improve connection time we should `forceRelay` or `natTraversalDisabled` ? If not, what are high level changes that helped improved connection time ?

We use AWS KVS to stream video from our IoT linux camera to iOS Mobile App. We have a terrible connection failure rate and connection setup time . We have been trying to improve `TTFF` but out of luck.

I tried to go through app code to find any obvious difference our implementation vs recommended in this application, But couldn't find any. It would be helpful if you could help list high level steps suggested to improve connection time and connection rate. 

We have 10s connection timeout and `trickleICE` enabled. Stream resolution is SD quality ( 640x480, H264, 600 kbps CBR). Our device and customers are in `US-west` region only. I can share more specific details about our implementation if needed.

Questions

1. Should we always force Relay initially to speed up connection time ?
2. Should we also reduce number of `TURN` server used + Relay candidates collected ?
3. Is it possible to achieve < 1s for majority of connections ( p90 / p75 )? considering IoT linux devices have good internet speed ( through PoE ) and Mobile App runs on any possible iPhone with data plan / WiFi connectivity.

**Comment by binbin821** [comment]
@darshan-verkada The following are the answers to your questions.
1. Should we always force Relay initially to speed up connection time?   
- No, we don't force Relay 

2. Should we also reduce number of `TURN` server used + Relay candidates collected ?   
- No, we use the default number of `TURN` server used + Relay candidates collected

3. Is it possible to achieve < 1s for majority of connections ( p90 / p75 )? considering IoT linux devices have good internet speed ( through PoE ) and Mobile App runs on any possible iPhone with data plan / WiFi connectivity. 
- I'm not sure how you define the connection time. In our project, we can collect the `preConnection time cost`(the total time on initiating signaling server connection) and `streaming time cost`(the total time from starting peer connection to playing first frame).

**Comment by binbin821** [comment]
@darshan-verkada The high level steps to improve connection time and connection rate mainly include three aspects.
1. Initialize pre-connection process when users cold start app
- Pre-connection is to setup a connection with KVS signaling server in advance.
2. Reuse previous active KVS signaling server connection
- For best performance, we only need to create a new KVS signaling connection when the connection is closed or there is an error received from the signaling service.
```js
// Once the signaling channel connection is closed, open signaling connection
if (!isSignalingClientOpen) {
    viewer.signalingClient.open();
} else {
    // Once the signaling channel connection is open, open peer to peer connection
    await startPeerConnect();
}
```
3. Bypass DNS and refresh the list of IceServer before expired
- The URIs in [IceServer](https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/API_signaling_IceServer.html) provide the different addresses and/or protocols that can be used to reach the TURN server. The form is as the following.

```js
"turns:52-80-94-187.t-89e91582.kinesisvideo.cn-north-1.amazonaws.com.cn:443?transport=udp",
"turns:52-80-94-187.t-89e91582.kinesisvideo.cn-north-1.amazonaws.com.cn:443?transport=tcp"
```

- In order to solve DNS failure and save the time to convert a hostname (such as 52-80-94-187.t-89e91582.kinesisvideo.cn-north-1.amazonaws.com.cn) into a computer-friendly IP address (such as 52.80.94.187),  we figure out a method to extract the IP address directly from the information contained in the domain, as shown in the following example.

```js
// URIs before extracting
"turns:52-80-94-187.t-89e91582.kinesisvideo.cn-north-1.amazonaws.com.cn:443?transport=udp",
"turns:52-80-94-187.t-89e91582.kinesisvideo.cn-north-1.amazonaws.com.cn:443?transport=tcp"

// URIS after extracting
"turns:52.80.94.187:443?transport=udp",
"turns:52.80.94.187:443?transport=tcp"
```

For more details, you can refer to [viewer.js](https://github.com/aws-samples/kvs-connection-time-optimization-for-webrtc-sample/blob/main/MobileApp/app/src/main/assets/kvs-webrtc/viewer.js)
